crates/ra_hir/src/ty.rs

```diff
@@ -872,6 +873,155 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
         }
     }
 
+    fn resolve_fields(&self, path: Option<&Path>) -> Option<(Ty, Vec<StructField>)> {
+        let def = path
+            .and_then(|path| self.module.resolve_path(self.db, &path).take_types())
+            .map(|def_id| def_id.resolve(self.db));
+
+        let def = if let Some(def) = def {
+            def
+        } else {
+            return None;
+        };
+
+        match def {
+            Def::Struct(s) => {
+                let fields: Vec<_> = self
+                    .db
+                    .struct_data(s.def_id())
+                    .variant_data
+                    .fields()
+                    .iter()
+                    .cloned()
+                    .collect();
+                Some((type_for_struct(self.db, s), fields))
+            }
+            Def::EnumVariant(ev) => {
+                let fields: Vec<_> = ev.variant_data(self.db).fields().iter().cloned().collect();
+                Some((type_for_enum_variant(self.db, ev), fields))
+            }
+            _ => None,
+        }
+    }
+
+    fn infer_tuple_struct(&mut self, path: Option<&Path>, subpats: &[PatId]) -> Ty {
+        let (ty, fields) = if let Some(x) = self.resolve_fields(path) {
```

Comment:
this will be really weird for non-tuple structs, but we have to still really implement tuple structs in the expression case anyway, so I guess it's fine for now